File: hooks/scripts/pre/validate-input.py
#!/usr/bin/env python3
from __future__ import annotations
import os
import sys

sys.path.insert(0, os.path.dirname(__file__))
from _lib import read_stdin, deny, passthrough


def validate_write(tool_input: dict) -> str | None:
    fp = tool_input.get("file_path", "")
    if not fp or not fp.startswith("/"):
        return "file_path must be an absolute path"
    content = tool_input.get("content", "")
    if not content:
        return "content must be non-empty"
    return None


def validate_edit(tool_input: dict) -> str | None:
    fp = tool_input.get("file_path", "")
    if not fp or not fp.startswith("/"):
        return "file_path must be an absolute path"
    old = tool_input.get("old_string", "")
    new = tool_input.get("new_string", "")
    if not old:
        return "old_string must be non-empty"
    if not new:
        return "new_string must be non-empty"
    if old == new:
        return "old_string and new_string must be different"
    return None


def validate_bash(tool_input: dict) -> str | None:
    cmd = tool_input.get("command", "")
    if not cmd or not cmd.strip():
        return "command must be non-empty"
    return None


def validate_read(tool_input: dict) -> str | None:
    fp = tool_input.get("file_path", "")
    if not fp or not fp.startswith("/"):
        return "file_path must be an absolute path"
    return None


def validate_webfetch(tool_input: dict) -> str | None:
    url = tool_input.get("url", "")
    if not url:
        return "url must be non-empty"
    if not url.startswith(("http://", "https://")):
        return "url must start with http:// or https://"
    return None


VALIDATORS = {
    "Write": validate_write,
    "Edit": validate_edit,
    "Bash": validate_bash,
    "Read": validate_read,
    "WebFetch": validate_webfetch,
}


def main() -> None:
    data = read_stdin()
    if not data:
        passthrough()

    tool_name = data.get("tool_name", "")
    tool_input = data.get("tool_input", {})

    validator = VALIDATORS.get(tool_name)
    if not validator:
        passthrough()

    error = validator(tool_input)
    if error:
        deny(f"[schema] {tool_name}: {error}")

    passthrough()


if __name__ == "__main__":
    main()

File: claude/hooks/scripts/pre/validate-input.mjs
#!/usr/bin/env node
import "../suppress-stderr.mjs";
import { deny, passthrough, readStdin } from "../_lib.mjs";

function validateWrite(toolInput) {
	const fp = toolInput.file_path ?? "";
	if (!fp?.startsWith("/")) return "file_path must be an absolute path";
	const content = toolInput.content ?? "";
	if (!content) return "content must be non-empty";
	return null;
}

function validateEdit(toolInput) {
	const fp = toolInput.file_path ?? "";
	if (!fp?.startsWith("/")) return "file_path must be an absolute path";
	const oldStr = toolInput.old_string ?? "";
	const newStr = toolInput.new_string ?? "";
	if (!oldStr) return "old_string must be non-empty";
	if (!newStr) return "new_string must be non-empty";
	if (oldStr === newStr) return "old_string and new_string must be different";
	return null;
}

function validateBash(toolInput) {
	const cmd = toolInput.command ?? "";
	if (!cmd?.trim()) return "command must be non-empty";
	return null;
}

function validateRead(toolInput) {
	const fp = toolInput.file_path ?? "";
	if (!fp?.startsWith("/")) return "file_path must be an absolute path";
	return null;
}

function validateWebFetch(toolInput) {
	const url = toolInput.url ?? "";
	if (!url) return "url must be non-empty";
	if (!url.startsWith("http://") && !url.startsWith("https://")) {
		return "url must start with http:// or https://";
	}
	return null;
}

const VALIDATORS = {
	Write: validateWrite,
	Edit: validateEdit,
	Bash: validateBash,
	Read: validateRead,
	WebFetch: validateWebFetch,
};

(async () => {
	try {
		const data = await readStdin();
		if (!data) passthrough();

		const toolName = data.tool_name ?? "";
		const toolInput = data.tool_input ?? {};

		const validator = VALIDATORS[toolName];
		if (!validator) passthrough();

		const error = validator(toolInput);
		if (error) deny(`[schema] ${toolName}: ${error}`);

		passthrough();
	} catch {
		passthrough();
	}
})();
